x/tradeshield/types/messages_pending_perpetual_order.go

```diff
@@ -6,84 +6,237 @@ import (
 )
 
 const (
-	TypeMsgCreatePendingPerpetualOrder = "create_pending_perpetual_order"
-	TypeMsgUpdatePendingPerpetualOrder = "update_pending_perpetual_order"
-	TypeMsgCancelPerpetualOrders       = "cancel_perpetual_order"
+	TypeMsgCreatePerpetualOpenOrder  = "create_perpetual_open_order"
+	TypeMsgCreatePerpetualCloseOrder = "create_perpetual_close_order"
+	TypeMsgUpdatePerpetualOrder      = "update_perpetual_order"
+	TypeMsgCancelPerpetualOrder      = "cancel_perpetual_order"
+	TypeMsgCancelPerpetualOrders     = "cancel_perpetual_orders"
 )
 
-var _ sdk.Msg = &MsgCreatePendingPerpetualOrder{}
+var _ sdk.Msg = &MsgCreatePerpetualOpenOrder{}
 
-func NewMsgCreatePendingPerpetualOrder(creator string) *MsgCreatePendingPerpetualOrder {
-	return &MsgCreatePendingPerpetualOrder{
-		OwnerAddress: creator,
+func NewMsgCreatePerpetualOpenOrder(
+	ownerAddress string,
+	orderType PerpetualOrderType,
+	triggerPrice TriggerPrice,
+	collateral sdk.Coin,
+	tradingAsset string,
+	position PerpetualPosition,
+	leverage sdk.Dec,
+	takeProfitPrice sdk.Dec,
+	stopLossPrice sdk.Dec,
+	poolId uint64,
+) *MsgCreatePerpetualOpenOrder {
+	return &MsgCreatePerpetualOpenOrder{
+		TriggerPrice:    &triggerPrice,
+		Collateral:      collateral,
+		OwnerAddress:    ownerAddress,
+		TradingAsset:    tradingAsset,
+		Position:        position,
+		Leverage:        leverage,
+		TakeProfitPrice: takeProfitPrice,
+		StopLossPrice:   stopLossPrice,
+		PoolId:          poolId,
+	}
+}
+
+func (msg *MsgCreatePerpetualOpenOrder) Route() string {
+	return RouterKey
+}
+
+func (msg *MsgCreatePerpetualOpenOrder) Type() string {
+	return TypeMsgCreatePerpetualOpenOrder
+}
+
+func (msg *MsgCreatePerpetualOpenOrder) GetSigners() []sdk.AccAddress {
+	creator, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
+	if err != nil {
+		panic(err)
+	}
+	return []sdk.AccAddress{creator}
+}
+
+func (msg *MsgCreatePerpetualOpenOrder) GetSignBytes() []byte {
+	bz := ModuleCdc.MustMarshalJSON(msg)
+	return sdk.MustSortJSON(bz)
+}
+
+func (msg *MsgCreatePerpetualOpenOrder) ValidateBasic() error {
+	_, err := sdk.AccAddressFromBech32(msg.OwnerAddress)
+	if err != nil {
+		return sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid owner address (%s)", err)
+	}
+
+	// Validate trigger price
+	if msg.TriggerPrice == nil {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "trigger price cannot be nil")
+	}
+
+	// Validate collateral
+	if !msg.Collateral.IsValid() {
+		return sdkerrors.Wrap(sdkerrors.ErrInvalidRequest, "invalid collateral")
+	}
+
+	// Validate trading asset
+	if msg.TradingAsset == "" {
```

Comment:
use sdk.ValidateDenom to validate denoms. there are length constraints and what characters allowed